crates/peer_binary_protocol/src/extended/ut_pex.rs

```diff
@@ -0,0 +1,143 @@
+
+#[derive(Debug)]
+pub struct PexPeerInfo {
+    pub flags: u8,
+    pub addr: SocketAddr,
+}
+
+impl PexPeerInfo {
+    pub fn from_bytes(buf: &[u8], flags: Option<u8>) -> anyhow::Result<Self> {
+        let (ip, port) = match buf.len() {
+            6 => {
+                let ip_bytes: &[u8; 4] = (&buf[0..4]).try_into()?;
+                let ip = IpAddr::from(*ip_bytes);
+                let port = BE::read_u16(&buf[4..6]);
+                (ip, port)
+            }
+            18 => {
+                let ip_bytes: &[u8; 16] = (&buf[0..16]).try_into()?;
+                let ip = IpAddr::from(*ip_bytes);
+                let port = BE::read_u16(&buf[16..18]);
+                (ip, port)
+            }
+            _ => anyhow::bail!("invalid pex peer info"),
+        };
+        Ok(Self {
+            flags: flags.unwrap_or(0),
+            addr: (ip, port).into(),
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Default, Deserialize)]
+pub struct UtPex<B> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    added: Option<B>,
+    #[serde(rename = "added.f")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    added_f: Option<B>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    added6: Option<B>,
+    #[serde(rename = "added6.f")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    added6_f: Option<B>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    dropped: Option<B>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    dropped6: Option<B>,
+}
+
+impl<B> CloneToOwned for UtPex<B>
+where
+    B: CloneToOwned,
+{
+    type Target = UtPex<<B as CloneToOwned>::Target>;
+    fn clone_to_owned(&self, within_buffer: Option<&Bytes>) -> Self::Target {
+        UtPex {
+            added: self.added.clone_to_owned(within_buffer),
+            added_f: self.added_f.clone_to_owned(within_buffer),
+            added6: self.added6.clone_to_owned(within_buffer),
+            added6_f: self.added6_f.clone_to_owned(within_buffer),
+            dropped: self.dropped.clone_to_owned(within_buffer),
+            dropped6: self.dropped6.clone_to_owned(within_buffer),
+        }
+    }
+}
+
+pub struct PeerAddrIterator<'a> {
+    addrs: &'a [u8],
+    flags: &'a [u8],
+    offset: usize,
+}
+
+impl<'a> Iterator for PeerAddrIterator<'a> {
+    type Item = PexPeerInfo;
+    fn next(&mut self) -> Option<Self::Item> {
+        const ADDR_SIZE: usize = 6;
+        if self.offset*ADDR_SIZE >= self.addrs.len() {
+            return None;
+        }
+
+        let addr = &self.addrs[self.offset*ADDR_SIZE..(self.offset+1)*ADDR_SIZE];
+        let flags = self.flags.get(self.offset);
+        self.offset += 1;
+        Some(PexPeerInfo::from_bytes(addr, flags.cloned()).unwrap()) // safe to unwrap as we assure slice length
+
+
+    }
+}
+
+impl<B> UtPex<B>
+where
+    B: AsRef<[u8]>,
+{
+    pub fn added_peers<'a>(&'a self) -> anyhow::Result<Box<dyn Iterator<Item = PexPeerInfo> + 'a>> {
```

Comment:
I'd avoid allocation here as it's quite easy to do.

two options: 
1. either use itertools::Either (minimal changes to your existing code, good enough)
2. Option is also an iterator, and you can flatten it.

Option 2 would be smth like this:

```
if self.added.as_ref().map(|added| added.len() % 6 != 0).unwrap_or(false) {
    bail!(...)
}

let ips = self.added.as_ref().map(|addrs| addrs.chunk_exact(6)).flatten();
let flags = self.added_f.as_ref().map(|flags| flags.iter().copied()).flatten();

return ips.zip(flags).filter_map(|(ip, flags)| PexPeerInfo::from_bytes(ip, flags).ok())

```

Reply:
added flags - added_f - are optional - so we can have addresses  but not flags, in that case zip will not work. But chunks  might make things easier - will look into it.

Reply:
Look into the PR I put into your repo I've done it already